examples/contrib/mtls.py

```diff
@@ -0,0 +1,194 @@
+            default="",
+        )
+        # Must be lazy. This makes the mitmproxy extract the client certificate
+        # before connecting to the server.
+        ctx.options.connection_strategy = "lazy"
+        # If the real server do not care about the host header, enable it to make it a bit faster.
+        ctx.options.keep_host_header = True
+
+    def configure(self, updated: set[str]):
+        # Override original process of loading certs.
+
+        if ctx.options.client_ca == "":
+            raise exceptions.OptionsError("client_ca is empty")
+        if ctx.options.server_ca == "":
+            raise exceptions.OptionsError("server_ca is empty")
+
+        if "client_ca" in updated:
+            ca_path = os.path.expanduser(ctx.options.client_ca)
+            self.clientCertStore = certs.CertStore.from_files(
+                ca_file=Path(ca_path),
+                dhparam_file=Path(ca_path + ".dhparam.pem"),
+            )
+        if "server_ca" in updated:
+            ca_path = os.path.expanduser(ctx.options.server_ca)
+            self.certstore = certs.CertStore.from_files(
+                ca_file=Path(ca_path),
+                dhparam_file=Path(ca_path + ".dhparam.pem"),
+            )
+            ctx.options.ssl_verify_upstream_trusted_ca = ctx.options.server_ca
+
+    def tls_start_client(self, data: tls.TlsData):
+        # In this stage, mitmproxy generates a fake cert to impersonate the real server.
+
+        def cb(_data: tls.TlsData):
+            # In addons/tlsconfig.py, request_client_cert is set to False.
+            # But we must verify the peer to get client certificates.
+            _ctx = _data.ssl_conn.get_context()
+            _ctx.set_verify(net.tls.Verify.VERIFY_PEER.value, net.tls.accept_all)
+
+        self.tls_start_client_cb = cb
+
+        super().tls_start_client(data)
+
+        server_cert = data.ssl_conn.get_certificate()
+        logging.info(
+            "tls_start_client: fake server cert: %s", server_cert.get_subject()
+        )
+
+    def tls_start_server(self, data: tls.TlsData):
+        # In this stage, we use the fake client cert to connect the server.
+
+        client_certs = data.context.client.certificate_list
+        if client_certs and len(client_certs) > 0:
+            c = client_certs[0]
+            entry = self.clientCertStore.get_cert(c.cn, [], c.organization, True)
+            logging.info(
+                "tls_start_server: client cert: CN=%s O=%s", c.cn, c.organization
+            )
+
+            def cb(_data: tls.TlsData):
+                # In addons/tlsconfig.py, client certs are provided by ctx.options.client_certs.
+                # But here use the cert in the certstore.
+                _ctx = _data.ssl_conn.get_context()
+                _ctx.use_privatekey(crypto.PKey.from_cryptography_key(entry.privatekey))
+                _ctx.use_certificate(entry.cert.to_pyopenssl())
```

Comment:
I haven't tested this, but can you set the store cert after `super().tls_start_server(data)`?

Reply:
Unfortunately, it does not take effect to change the context if `tls_start.ssl_conn.set_connect_state()` has been invoked. Maybe we can do monkey patch just like [check_ssl_pinning.py](https://github.com/mitmproxy/mitmproxy/blob/10.1.1/examples/contrib/check_ssl_pinning.py#L94) do.